lib/publiccloud/instance.pm

```diff
@@ -333,73 +333,143 @@ sub wait_for_guestregister {
-                die "ssh permission denied (pubkey)" unless $args{ignore_wrong_pubkey};
-            }
-        }
-        sleep $sleep_period;
-    }
+    my $instance_msg = "instance: $self->{instance_id}, public IP: $self->{public_ip}";
+    my ($duration, $exit_code, $sshout, $sysout);
 
-    script_run("ssh  -i /root/.ssh/id_rsa $args{username}\@$args{public_ip} true", timeout => 360);
-    # Debug output: We have occasional error in 'journalctl -b' - see poo#96464 - this will be removed soon.
-    # Exclude 'mr_test/saptune' test case as it will introduce random softreboot failures.
-    if (!get_var('PUBLIC_CLOUD_SLES4SAP')) {
-        $self->run_ssh_command(cmd => 'sudo journalctl -b', proceed_on_failure => 1, username => $args{username}, timeout => 360);
-    }
+    # Looping until SSH port 22 is reachable or timeout.
+    while (($duration = time() - $start_time) < $args{timeout}) {
+        $exit_code = script_run('nc -vz -w 1 ' . $self->{public_ip} . ' 22', quiet => 1);
+        last if (isok($exit_code) and not $args{wait_stop});    # ssh port open ok
+        last if (not isok($exit_code) and $args{wait_stop});    # ssh port closed ok
+        sleep 1;
+    }    # endloop
 
-    unless ($args{proceed_on_failure}) {
-        my $error_msg;
-        if ($check_port) {
-            $error_msg = sprintf("Unable to reach SSH port of instance %s with public IP:%s within %d seconds", $self->{instance_id}, $self->{public_ip},
-                $args{timeout});
-        }
-        else {
-            $error_msg = sprintf("Can not reach systemd target on instance %s with public IP:%s within %d seconds",
-                $self->{instance_id}, $self->{public_ip}, $args{timeout});
-        }
-        croak($error_msg);
+    # exit_code is 0 when shell script is ok
+    if (isok($exit_code)) {
+        $sshout = "SSH port is open\n";
     }
+    else {
+        $sshout = "SSH port is not open failed access\n";
+        $sshout .= "as expected by stopping: OK.\n" if $args{wait_stop};
+    }    # endif
+
+    # check also remote system is running, looping by --wait option:
+    if ($args{systemup_check} and isok($exit_code)) {
+        script_run("ssh-keyscan $args{public_ip} | tee -a ~/.ssh/known_hosts")
+          if (get_var('PUBLIC_CLOUD_PERF_COLLECT') or get_var('PUBLIC_CLOUD_CHECK_BOOT_TIME') or get_var('PUBLIC_CLOUD_CHECK_REBOOT'));
+        while (($duration = time() - $start_time) < $args{timeout}) {
+            # On boottime test we do hard reboot which may change the instance address:
+            # timeout recalculated removing consumed time until now
+            $sysout = $self->ssh_script_output(cmd => 'sudo systemctl is-system-running',
+                timeout => $args{timeout} - $duration, proceed_on_failure => 1, username => $args{username});
+            # result check
+            if ($sysout =~ m/Permission denied \(publickey\).*/) {
+                $exit_code = 2;
```

Comment:
Any reason for the different exit codes (0, 1, 2, undef)? They're set, but never used save by `isok()` which only checks for defined and 0.

Reply:
It is basically to allow to distinguish that status error in the loop, 
because 2 occurrs for `pubkey` issue, particular case (conditioned if `ignore` is true or die) that could change state after some retries; 
similar for `undef` that is not really an issue, but a transitory status expecting changes after retries (unconditioned, never die).

Reply:
Indeed after this clarification , the `else` status could be 2 too :) (because allowed to retry)